examples/inspection/plot_permutation_importance_multicollinear.py

```diff
@@ -0,0 +1,122 @@
+# training.
+perm_importance = permutation_importance(clf, X_train, y_train, n_rounds=10,
+                                         random_state=42)
+perm_sorted_idx = np.mean(perm_importance, axis=-1).argsort()
+
+tree_importance_sorted_idx = np.argsort(clf.feature_importances_)
+tree_indicies = np.arange(1, len(clf.feature_importances_)+1)
+
+fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(12, 8))
+ax1.barh(tree_indicies, clf.feature_importances_[tree_importance_sorted_idx])
+ax1.set_yticklabels(data.feature_names)
+ax1.set_yticks(tree_indicies)
+ax2.boxplot(perm_importance[perm_sorted_idx].T, vert=False,
+            labels=data.feature_names)
+fig.tight_layout()
+plt.show()
+
+##############################################################################
+# Handling Multicollinear Features
+# --------------------------------
+# When features are collinear, permutating one feature will have little
+# effect on the models performance because it can get the same information
+# from a correlated feature. One way to handle multicollinear features is by
+# performing hierarchical clustering on the Spearman rank-order correlations,
+# picking a threshold, and keeping a single feature from each cluster. First,
+# we plot a heatmap of the correlated features:
+corr = spearmanr(X).correlation
+corr_linkage = hierarchy.ward(corr)
+dendro = hierarchy.dendrogram(corr_linkage, no_plot=True)
+dendro_order = np.array(dendro['ivl'], dtype='int')
+dendro_idx = np.arange(0, len(dendro_order))
+
+fig, ax = plt.subplots(figsize=(12, 8))
+ax.imshow(corr[dendro_order, :][:, dendro_order])
+ax.set_xticks(dendro_idx)
+ax.set_yticks(dendro_idx)
+ax.set_xticklabels(data.feature_names[dendro_order], rotation='vertical')
+ax.set_yticklabels(data.feature_names[dendro_order])
+plt.show()
+
+##############################################################################
+# Next, we pick a threshold to group our features into clusters and choose a
+# feature from each cluster to keep, select those features from our dataset,
+# and train a new random forest. The test accuracy of the new random forest did
+# not changed much compared to the random forest trained on the complete
+# dataset.
+cluster_ids = hierarchy.fcluster(corr_linkage, 1, criterion='distance')
+cluster_id_to_feature_ids = defaultdict(list)
+for idx, cluster_id in enumerate(cluster_ids):
+    cluster_id_to_feature_ids[cluster_id].append(idx)
+selected_features = [v[0] for v in cluster_id_to_feature_ids.values()]
+
+X_train_sel = X_train[:, selected_features]
+X_test_sel = X_test[:, selected_features]
+
+clf_sel = RandomForestClassifier(n_estimators=100, random_state=42)
+clf_sel.fit(X_train_sel, y_train)
+print("Accuracy on test data with features removed: {:.2f}".format(
+      clf_sel.score(X_test_sel, y_test)))
+
+##############################################################################
+# Lastly, we plot the permutation importance with new random forest on the
+# test dataset.
```

Comment:
Why not also plot the feature importance computed directly from the trees (i.e. replicating the previous plot)?

Reply:
I wanted to focus on the multicollinear aspect for this example. The other examples goes into detail when looking at tree based feature importances.

Reply:
I think it's still interesting to be able to compare how the tree-based FI behaves when correlated features are removed, since the beginning of the examples starts off as a comparison.

Won't push it further though

Reply:
Hmm okay, I'll see if I can fit it into the narrative.